File: examples/chatapp/python-be/app/gql/ops.py
from __future__ import annotations

import forge_py
import strawberry
from strawberry.types import Info

from .helpers import FAIL_QUEUE, current_user, map_forge, require_admin, require_user
from .types import OpsStats


@strawberry.type
class OpsQuery:
    @strawberry.field(
        description="Whether the `reactions_v2` feature flag is enabled for the current user"
        " (forge config)."
    )
    async def reactions_enabled(self, info: Info) -> bool:
        u = await current_user(info)
        if u is None:
            return False
        return await info.context["forge"].flag("reactions_v2", False, str(u["id"]))

    @strawberry.field(
        description="Developer-tools gauges (kv scan + DLQ depth) for the settings page."
    )
    async def ops_stats(self, info: Info) -> OpsStats:
        await require_user(info)
        forge = info.context["forge"]
        try:
            online = len(await forge.kv_scan("online:", 1000))
        except forge_py.ForgeError:
            online = 0
        depth = await forge.queue_depth(f"{FAIL_QUEUE}.dlq")
        dlq_count = depth.visible + depth.in_flight + depth.delayed
        return OpsStats(online_count=online, dlq_count=dlq_count)


@strawberry.type
class OpsMutation:
    @strawberry.mutation(
        description="Set the `reactions_v2` feature-flag rollout percentage (forge config)."
    )
    async def set_reactions_rollout(self, info: Info, percent: int) -> bool:
        await require_admin(info)
        pct = max(0, min(100, percent))
        try:
            await info.context["forge"].set_flag_percent("reactions_v2", pct)
        except forge_py.ForgeError as e:
            raise map_forge(e) from e
        return True

    @strawberry.mutation(
        description="Enqueue a job destined to dead-letter (forge queue DLQ demo)."
    )
    async def trigger_failing_job(self, info: Info) -> bool:
        await require_admin(info)
        try:
            await info.context["forge"].queue_enqueue(FAIL_QUEUE, "boom")
        except forge_py.ForgeError as e:
            raise map_forge(e) from e
        return True

File: examples/chatapp/rust-be/src/gql/ops.rs
use async_graphql::{Context, Object, Result};

use crate::context::CurrentUser;
use crate::gql::helpers::{app, map_db, me, require_admin};
use crate::gql::types::OpsStats;

#[derive(Default)]
pub struct OpsQuery;

#[Object]
impl OpsQuery {
    async fn reactions_enabled(&self, ctx: &Context<'_>) -> Result<bool> {
        let Some(user) = ctx.data_opt::<CurrentUser>() else {
            return Ok(false);
        };
        Ok(app(ctx)?.reactions_enabled(user.id).await)
    }

    async fn ops_stats(&self, ctx: &Context<'_>) -> Result<OpsStats> {
        let c = app(ctx)?;
        me(ctx)?;
        Ok(OpsStats {
            online_count: c.online_count().await as i32,
            dlq_count: c.dlq_count().await as i32,
        })
    }
}

#[derive(Default)]
pub struct OpsMutation;

#[Object]
impl OpsMutation {
    async fn set_reactions_rollout(&self, ctx: &Context<'_>, percent: i32) -> Result<bool> {
        let c = app(ctx)?;
        require_admin(ctx)?;
        c.set_reactions_rollout(percent.clamp(0, 100) as u8)
            .await
            .map_err(map_db)?;
        Ok(true)
    }

    async fn trigger_failing_job(&self, ctx: &Context<'_>) -> Result<bool> {
        let c = app(ctx)?;
        require_admin(ctx)?;
        c.enqueue_failing().await.map_err(map_db)?;
        Ok(true)
    }
}
